### PR TITLE
Fix captured random state in compile

### LINKED ISSUE #1439: temperature > 0 is silently deterministic on the second call onward — categorical_sampling's @mx.compile(outputs=mx.random.state) doesn't advance the PRNG
Under inference with mlx-lm (and mlx-vlm, which reuses
`mlx_lm.sample_utils`), `temperature > 0` produces **byte-identical
output across repeat calls within a single process**, regardless of
`mx.random.seed()`, the `seed=` kwarg to `generate_step`, or any
sampling parameters.

Sampling appears to work for the first ~2 token positions in a process,
then silently collapses to a deterministic "stuck" sampler. Each token
position thereafter selects whichever token contains the same fixed
percentile of its distribution — confident decisions look greedy, less
confident decisions look weirdly off-distribution.

Server bounces are the only way to change output. Different processes
get different "stuck" sampler states, so behavior varies across
restarts but never within a session.

## Versions

- mlx-lm 0.31.3
- mlx-vlm 0.6.3 (also affected — reuses `mlx_lm.sample_utils`)
- mlx 0.31.2
- Python 3.12, macOS Darwin, Apple Silicon M-series

## Evidence

Two `/v1/chat/completions` calls, same prompt, `temperature=1.0`:

```
Request 1:  md5(content) = 7ac0dba2  "The user wants a haiku about coffee..."
Request 2:  md5(content) = 7ac0dba2  "The user wants a haiku about coffee..."
            ^^^^^^^^^^^^^ byte-identical
```

Tracing `mx.random.state` at each `categorical_sampling` call inside
the server shows the state advances once, then freezes:

```
call 1 state: [array([0, 2129127115])]                  ← initial
call 2 state: [array([1131398383, 4059925319])]         ← advanced
call 3 state: [array([1131398383, 4059925319])]         ← stuck
call 4 state: [array([1131398383, 4059925319])]         ← stuck
call 5 state: [array([1131398383, 4059925319])]         ← stuck
call 6 state: [array([1131398383, 4059925319])]         ← stuck
```

Same calls in a standalone Python script (no inference, just
`categorical_sampling(fake_logits, 1.0)` in a loop) DO advance state
correctly — so the bug only manifests in a real generation pipeline.
Something about `generate_step`'s context (compiled forward pass,
multiple `mx.compile`-decorated helpers sharing state, `mx.stream`
nesting, possibly thread affinity in a single-worker threadpool)
triggers the freeze.

## Workaround (confirmed working)

Replace `categorical_sampling` with a non-compiled version at process
startup:

\`\`\`python
import mlx.core as mx
import mlx_lm.sample_utils as su

def _cat_uncompiled(logits, temp):
    return mx.random.categorical(logits * (1 / temp))

su.categorical_sampling = _cat_uncompiled

# mlx-vlm imports the same symbol — patch it too if used:
try:
    import mlx_vlm.sample_utils as vsu
    vsu.categorical_sampling = _cat_uncompiled
except ImportError:
    pass
\`\`\`

After this patch:
- Same prompt + same `temperature > 0` → varied output across calls ✓
- `mx.random.seed()` actually changes output ✓
- Reasoning models with thinking blocks now produce varied actions, not
  just varied reasoning text

## Why this matters

CoT / thinking-mode models are particularly sensitive. They produce
varied `reasoning_content` (the thinking-phase sampler hits a
different code path) but byte-identical action/content output. So
the model "thinks" differently each turn while always emitting the
same final answer — sampling-driven exploration is silently lost.

For tool-calling agents this manifests as unbreakable loops: the model
re-thinks its decision each turn but always emits the same broken tool
call because the action distribution is peaked and the stuck percentile
always lands on the same top token.

For standard chat use, output across sessions feels "inconsistent"
(each process restart picks a different stuck state) while output
within a session is suspiciously deterministic. Easy to misattribute
to model variance or quantization noise.

## Suggested fix

The most direct option: remove the `@partial(mx.compile, ...)`
decorator from `categorical_sampling`. The body is a single
`mx.random.categorical` call — compile gains are negligible and the
correctness cost is total.

Longer-term: investigate whether `mx.compile` with
`outputs=mx.random.state` is supposed to bind state on first invocation
under heavy inference contexts. If yes, document the contract; if no,
that's an mlx-level issue worth filing separately.

## Related code

- `mlx_lm/sample_utils.py` — `categorical_sampling` decorator
- `mlx_lm/generate.py` (and `mlx_vlm/generate/`) — sampler call sites
  during decode loop
- `mlx_vlm/sample_utils.py` — re-exports the affected symbol

### PR DESCRIPTION
**Pull request by angeloskath**
This is an interesting bug. In our attempts at thread-safety, compile now captures the random state of the thread that created the function and not the thread that called the function which can (and often is) different.

There isn't too clean of a solution that I could think of so please be as strict as possible in your judgment. The current approach is twofold

1. Make the random state a singleton object that fetches the thread local state. This makes it ok to capture on any thread. Accessing its one element will get you the correct array based on which thread you are on.
2. The ugly part of this fix is that `tree_visit` and `tree_visit_update` now have a special case where if they encounter a random state they instead operate on the underlying array. This is ugly because it is not necessarily consistent with `tree_map` or multi-tree `tree_visit`.

This closes https://github.com/ml-explore/mlx-lm/pull/1444 and fixes https://github.com/ml-explore/mlx-lm/issues/1439 .

### CODE REVIEW

File: python/src/random.cpp [hunk 1]
```diff
@@ -63,15 +64,57 @@ PyKeySequence& default_key() {
...
 }
 
+// A process-global sentinel for `mx.random.state`. Since it is the same object
+// on every thread, capturing it (e.g. with `mx.compile`) is thread-independent;
+// the pytree traversal in trees.cpp resolves it to the calling thread's key.
+class RandomState {};
+
+nb::object random_state_sentinel() {
+  static nb::object sentinel;
+  if (sentinel.ptr() == nullptr) {
```
Comment (zcbenz) [inline python/src/random.cpp]: This is not thread safe.

File: python/src/trees.cpp [hunk 2]
```diff
@@ -153,7 +154,10 @@ void tree_visit(
 void tree_visit(nb::handle tree, std::function<void(nb::handle)> visitor) {
   std::function<void(nb::handle)> recurse;
   recurse = [&](nb::handle subtree) {
-    if (nb::isinstance<nb::list>(subtree) ||
+    if (is_random_state(subtree)) {
+      visitor(nb::cast(random_state_key()));
```
Comment (zcbenz) [inline python/src/trees.cpp:159]: I wonder if we can do this in a less intrusive way by replacing

https://github.com/ml-explore/mlx/blob/4367c73b60541ddd5a266ce4644fd93d20223b6e/python/src/transforms.cpp#L560-L563

with

```cpp
if (!captured_inputs.is_none()) { 
  trace_captures.push_back(random_state_sentinel());
  flat_in_captures.push_back(random_state_key());
  tree_replace(captured_inputs, trace_captures, flat_in_captures);
}
```

File: python/src/random.cpp (hunk 1 above)
Comment (angeloskath) [reply python/src/random.cpp]: Yeah true let me change it to a lambda call. I just kind of took it from the structure sentinel in trees.cpp, will change both.

File: python/src/trees.cpp (hunk 2 above)
Comment (angeloskath) [reply python/src/trees.cpp:159]: The problem is the random state can be in any place in the captured inputs. As a result we have two options

1. Walk the captured inputs and replace the random state with the thread local one before flattening
2. Do it when the random state is encountered while flattening and while writing back

Option 1 has the overhead of 2 extra tree walks which for say 100k nodes (like a big transformer or sth) can be non-negligible.

Option 2 is intrusive and ugly.

Comment (zcbenz) [reply python/src/trees.cpp:159]: That makes sense, we can probably add a callback to `tree_flatten` to make it cheap to do replacement but it feels over-engineered and still does not fix the hook in `tree_visit_update`, I'm good with current approach and I can't come with a better one.

**Review by zcbenz — APPROVED** [review]
